### PR TITLE
Don't allow duplicate Versions in the API (fixes #1025)

### LINKED ISSUE #1025: Don't create multiple Version objects for a version string
There are some cases in which the API may create multiple `Version` objects with the same version string. This causes problems with automated signing where there's a race condition and messes up the review queue when not in automated signing.

* [ ] Change the [`Version` creation check](https://github.com/mozilla/olympia/blob/77c7e3f0e514e165060495b2173b27820d9cadc0/apps/devhub/tasks.py#L71-L72) from "is this the most recent upload" to "does a `Version` already exist."
* [ ] Wrap the version check code in a transaction to ensure we are looking at a stable snapshot and avoid the quick submission race condition.

### PR DESCRIPTION
**Pull request by mstriemer**

### CODE REVIEW

File: apps/devhub/tasks.py [hunk 1]
```diff
@@ -67,16 +68,29 @@ def validate_and_submit(addon, file_, listed=None):
...
+    if file_.passed_all_validations:
+        create_version_for_upload(addon, file_)
+    else:
+        log.info('Skipping version creation for {file_id} that failed '
+                 'validation'.format(file_id=file_pk))
+
+
+@write
+@transaction.atomic
+def create_version_for_upload(addon, file_, sleep=None):
```
Comment (kumar303) [inline apps/devhub/tasks.py]: I don't think `sleep=None` is needed anymore

File: apps/devhub/tasks.py [hunk 2]
```diff
@@ -67,16 +68,29 @@ def validate_and_submit(addon, file_, listed=None):
...
+    if (addon.fileupload_set.filter(
+            created__gt=file_.created, version=file_.version).exists()
+            or addon.versions.filter(version=file_.version).exists()):
```
Comment (kumar303) [inline apps/devhub/tasks.py]: maybe this extra query is enough to fix the bug :beers:

File: apps/devhub/tasks.py [hunk 3]
```diff
@@ -67,16 +68,29 @@ def validate_and_submit(addon, file_, listed=None):
...
+    else:
+        log.info('Skipping version creation for {file_id} that failed '
+                 'validation'.format(file_id=file_pk))
+
+
+@write
+@transaction.atomic
+def create_version_for_upload(addon, file_, sleep=None):
+    if (addon.fileupload_set.filter(
+            created__gt=file_.created, version=file_.version).exists()
```
Comment (kumar303) [inline apps/devhub/tasks.py]: nit: this line should be indented once more to indicate that it's part of a nested clause and not part of the or statement below it

File: apps/devhub/tasks.py [hunk 4]
```diff
@@ -67,16 +68,29 @@ def validate_and_submit(addon, file_, listed=None):
 def submit_file(addon_pk, file_pk):
     addon = Addon.unfiltered.get(pk=addon_pk)
```
Comment (kumar303) [inline apps/devhub/tasks.py]: can we also do this select without caching? like `Addon.uncached` or something?

**Comment by kumar303** [comment]
I don't see us overriding the default isolation level in mysql so that might have an affect on the behavior you're seeing while trying to test this. Some info: http://blog.ionelmc.ro/2014/12/28/terrible-choices-mysql/#the-transaction-isolation-level

Also, you may have some tables in your local db that were created with syncdb as opposed to created from migrations. Syncdb might be creating non-innodb tables but hopefully not (I think django might try to create innodb tables by default).

**Comment by kumar303** [comment]
r+wc

I think it's worth adding a test for the new query you added: `addon.versions.filter(version=file_.version).exists()`

We should also make sure caching isn't involved in any of these queries, especially since we know that we still have cache invalidation problems in prod. I think you can wrap both functions in [@skip_cache](https://github.com/mozilla/olympia/blob/master/apps/amo/decorators.py#L145) or something.

File: apps/devhub/tasks.py (hunk 3 above)
Comment (mstriemer) [inline apps/devhub/tasks.py]: I agree. Interestingly my PEP8 linter complains about that. I switched it to line up with the `(`.

**Comment by mstriemer** [comment]
I changed my `settings_base.py` to have the `init_command` and this works as you would expect. I don't know if it is safe to make that change or not. It seems like what we'd want though.

```py
DATABASES['default']['OPTIONS'] = {
    'sql_mode': 'STRICT_ALL_TABLES',  # This was already being set.
    'init_command': 'SET SESSION TRANSACTION ISOLATION LEVEL READ COMMITTED',
}

**Comment by mstriemer** [comment]
I made the API reject uploads earlier also rather than accepting it and then waiting for the validation to complete to not create the `Version`. It's still quite possible that you get multiple `FileUpload`s with the same version string. That should be expected as if the validation fails we should still allow more uploads.

It looks like you need to set the `init_command` for each database so I'm assuming this won't get applied to databases other than `default`. I'm not sure how the databases are configured but since I know we have more than one this is likely going to get them out of sync in how they handle transactions.

Even without the transaction working the way we want this should help.

File: apps/devhub/tasks.py (hunk 4 above)
Comment (magopian) [inline apps/devhub/tasks.py]: We have [Addon.no_cache()](https://github.com/washort/django-cache-machine/blob/master/caching/base.py#L61) if that's what you're talking about?

**Pull request merged**